x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+            if (JWSAlgorithm.Family.HMAC_SHA.contains(requestedAlgorithm)) {
+                clientSecret = new Secret(rpConfig.getClientSecret().toString());
+                validator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm, clientSecret);
+            } else {
+                String jwkSetPath = opConfig.getJwkSetPath();
+                if (jwkSetPath.startsWith("https://")) {
+                    validator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm,
+                        new URL(jwkSetPath), privilegedResourceResolver);
+                } else {
+                    final Path path = realmConfig.env().configFile().resolve(jwkSetPath);
+                    final JWKSet jwkSet = JWKSet.load(path.toFile());
+                    validator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm, jwkSet);
+                }
+
+            }
+            JWTClaimsSet verifiedIdTokenClaims = validator.validate(idToken, expectedNonce).toJWTClaimsSet();
+            if (logger.isTraceEnabled()) {
+                logger.trace("Received the Id Token for the user: [{}]", verifiedIdTokenClaims);
+            }
+            return verifiedIdTokenClaims;
+
+        } catch (Exception e) {
+            logger.debug("Failed to parse or validate the ID Token. ", e);
+            throw new ElasticsearchSecurityException("Failed to parse or validate the ID Token", e);
+        } finally {
+            if (null != clientSecret) {
+                clientSecret.erase();
+            }
+        }
+    }
+
+    /**
+     * Validate that the response we received corresponds to the response type we requested
+     *
+     * @param response The {@link AuthenticationSuccessResponse} we received
+     * @throws ElasticsearchSecurityException if the response is not the expected one for the configured response type
+     */
+    private void validateResponseType(AuthenticationSuccessResponse response) {
+        if (rpConfig.getResponseType().equals(response.impliedResponseType()) == false) {
+            logger.debug("Unexpected response type [{}], while [{}] is configured", response.impliedResponseType(),
+                rpConfig.getResponseType());
+            throw new ElasticsearchSecurityException("Received a response with an unexpected response type");
+        }
+    }
+
+    /**
+     * Validate that the state parameter the response contained corresponds to the one that we generated in the
+     * beginning of this authentication attempt and was stored with the user's session at the facilitator
+     *
+     * @param expectedState The state that was originally generated
+     * @param state         The state that was contained in the response
+     */
+    private void validateState(State expectedState, State state) {
+        if (null == state || null == expectedState) {
```

Comment:
Can you please expand this `if` ? I think it would be helpful during debugging to know which of the values is missing but expected.